proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1223 @@
+
+extension Foundation.Data {
+  var storage: Span<UInt8> { get }
+}
+
+extension String.UTF8View {
+  // note: this could borrow a temporary copy of the `String`'s storage
+  var storage: Span<Unicode.UTF8.CodeUnit> { _read }
+}
+extension Substring.UTF8View {
+  // note: this could borrow a temporary copy of the `Substring`'s storage
+  var storage: Span<Unicode.UTF8.CodeUnit> { _read }
+}
+extension Character.UTF8View {
+  // note: this could borrow a temporary copy of the `Character`'s storage
+  var storage: Span<Unicode.UTF8.CodeUnit> { _read }
+}
+
+extension SIMD {
+  var storage: Span<Self.Scalar> { get }
+}
+extension KeyValuePairs {
+  var storage: Span<(Self.Key, Self.Value)> { get }
+}
+extension CollectionOfOne {
+  var storage: Span<Element> { get }
+}
+
+extension Slice: ContiguousStorage where Base: ContiguousStorage {
+  var storage: Span<Base.Element> { get }
+}
+```
+
+In addition to the the safe types above gaining the `storage` property, the `UnsafeBufferPointer` family of types will also gain access to a `storage` property. This enables interoperability of `Span`-taking API. While a `Span` binding created from an `UnsafeBufferPointer` exists, the memory that underlies it must not be deinitialized or deallocated.
+
+```swift
+extension UnsafeBufferPointer {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<Element> { get }
+}
+extension UnsafeMutableBufferPointer {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<Element> { get }
+}
+extension UnsafeRawBufferPointer {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<UInt8> { get }
+}
+extension UnsafeMutableRawBufferPointer {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<UInt8> { get }
+}
+```
+
+#### Using `Span` with C functions or other unsafe code:
+
+`Span` has an unsafe hatch for use with unsafe code.
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  func withUnsafeBufferPointer<E: Error, Result: ~Copyable & ~Escapable>(
+    _ body: (_ buffer: UnsafeBufferPointer<Element>) throws(E) -> Result
+  ) throws(E) -> Result
+}
+
+extension Span where Element: BitwiseCopyable {
+  func withUnsafeBytes<E: Error, Result: ~Copyable & ~Escapable>(
+    _ body: (_ buffer: UnsafeRawBufferPointer) throws(E) -> Result
+  ) throws(E) -> Result
+}
+```
+
+### Complete `Span` API:
+
+```swift
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: UnsafePointer<Element>
+  internal var _count: Int
+}
+```
+
+##### Creating a `Span`:
+
+The initialization of a `Span` instance from an unsafe pointer is an unsafe operation. When it is initialized correctly, subsequent uses of the borrowed instance are safe. Typically these initializers will be used internally to a container's implementation of functions or computed properties that return a borrowed `Span`.
```

Comment:
I think it might be more correct to say that not only does `Span` have to be initialized correctly, but the memory backing the `Span` has to be managed correctly until the `Span` is no longer in use. For example, guaranteeing the lifetime. For example, one issue we have seen recently involves `mmap`ing some bytes, something unmapping it, then access to the now-gone values.

Reply:
I improved this paragraph with a more explicit requirement.